daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/ContractStateMachine.scala

```diff
@@ -324,6 +326,98 @@ class ContractStateMachine[Nid](mode: ContractKeyUniquenessMode) {
       case Nil => throw new IllegalStateException("Not inside a rollback scope")
       case _ :: tailStack => this.copy(rollbackStack = tailStack)
     }
+
+    private def withinRollbackScope: Boolean = rollbackStack.nonEmpty
+
+    /** Let `resolver` be a [[KeyResolver]] that can be used during interpretation to obtain a transaction `tx`
+      * in modes [[com.daml.lf.transaction.ContractKeyUniquenessMode.ContractByKeyUniquenessMode]],
+      * or `Map.empty` in mode [[com.daml.lf.transaction.ContractKeyUniquenessMode.Strict]].
+      * Let `this` state be the result of iterating over `tx` up to a node `n` exclusive using `resolver`.
+      * Let `substate` be the state obtained after fully iterating over the subtree rooted at `n` starting from [[State.empty]],
+      * using the resolver [[projectKeyResolver]](`resolver`)
+      * in modes [[com.daml.lf.transaction.ContractKeyUniquenessMode.ContractByKeyUniquenessMode]]
+      * or `Map.empty` in mode [[com.daml.lf.transaction.ContractKeyUniquenessMode.Strict]].
+      * Then, the returned state is the same as if
+      * the iteration over tx continued from `this` state through the whole subtree rooted at `n` using `resolver`.
+      *
+      * The iteration over the subtree rooted at `n` from `this` using the projected resolver fails
+      * if and only if advancing `this` using `substate` does, but the error may be different.
+      */
+    def advance(resolver: KeyResolver, substate: State): Either[KeyInputError, State] = {
+      require(
+        !substate.withinRollbackScope,
+        "Cannot lift a state over a substate with unfinished rollback scopes",
+      )
+
+      def keyMappingFor(key: GlobalKey): Option[KeyMapping] =
+        this.activeState.keys.get(key).orElse(this.globalKeyInputs.get(key).map(_.toKeyMapping))
+
+      // We want consistent key lookups within an action in any contract key mode.
+      def consistentGlobalKeyInputs: Either[KeyInputError, Unit] = {
+        substate.globalKeyInputs
+          .collectFirst {
+            case (key, KeyCreate)
+                if keyMappingFor(key).exists(_ != KeyInactive) &&
+                  mode != ContractKeyUniquenessMode.Off =>
+              DuplicateContractKey(key)
+            case (key, NegativeKeyLookup) if keyMappingFor(key).exists(_.isDefined) =>
+              InconsistentKeys(key)
+            case (key, Transaction.KeyActive(cid))
+                if keyMappingFor(key).exists(km => km != KeyActive(cid)) =>
+              InconsistentKeys(key)
+          }
+          .toLeft(())
+      }
+
+      for {
+        _ <- consistentGlobalKeyInputs
+      } yield {
+        val next = ActiveLedgerState(
+          consumedBy = this.activeState.consumedBy ++ substate.activeState.consumedBy,
```

Comment:
I wonder if in your planned usage of this method `this` and `substate` could consume the same contract id? And if so, if it is ok to overwrite the first archival by the second archival.

In any case, such a scenario should be tested. To make sure that advancing this over substate succeeds if and only if an iteration over the entire tx succeeds.

Reply:
As @cocreature pointed out, this is not meant to happen, although this may change when we introduce internal consistency checks in Canton (https://github.com/DACH-NY/canton/issues/713). I'll document this limitation for now.